ui/src/features/Editor/useCanvasCopyPaste.ts

```diff
@@ -221,46 +249,176 @@ export default ({
     [],
   );
 
+  const processNodesAndEdges = useCallback(
+    (selectedNodes: Node[], selectedEdges: Edge[]) => {
+      const processedNodeIds = new Set();
+      const nodesToProcess = [...selectedNodes];
+      const edgesToProcess = [...selectedEdges];
+
+      selectedNodes.forEach((node) => {
+        processedNodeIds.add(node.id);
+      });
+
+      const batchNodeIds = selectedNodes
+        .filter((node) => node.type === "batch")
+        .map((node) => node.id);
+
+      nodes.forEach((node) => {
+        if (
+          node.parentId &&
+          batchNodeIds.includes(node.parentId) &&
+          !processedNodeIds.has(node.id)
+        ) {
+          nodesToProcess.push(node);
+          processedNodeIds.add(node.id);
+        }
+      });
+
+      const allNodeIds = new Set(nodesToProcess.map((node) => node.id));
+
+      const processedEdgeIds = new Set(selectedEdges.map((edge) => edge.id));
+
+      edges.forEach((edge) => {
+        if (
+          (allNodeIds.has(edge.source) || allNodeIds.has(edge.target)) &&
+          !processedEdgeIds.has(edge.id)
+        ) {
+          edgesToProcess.push(edge);
+          processedEdgeIds.add(edge.id);
+        }
+      });
+
+      return {
+        nodes: nodesToProcess,
+        edges: edgesToProcess,
+      };
+    },
+    [nodes, edges],
+  );
+
   const handleCopy = useCallback(async () => {
     const selected: { nodes: Node[]; edges: Edge[] } | undefined = {
       nodes: nodes.filter((n) => n.selected),
       edges: edges.filter((e) => e.selected),
     };
     let referencedWorkflows: Workflow[] = [];
-    if (selected.nodes.some((n) => n.type === "reader"))
+
+    if (selected.nodes.length === 0 && selected.edges.length === 0) return;
+
+    const { nodes: nodesToCopy, edges: edgesToCopy } = processNodesAndEdges(
+      selected.nodes,
+      selected.edges,
+    );
+
+    if (nodesToCopy.some((n) => n.type === "subworkflow")) {
+      referencedWorkflows = collectSubworkflows(nodesToCopy, rawWorkflows);
+      if (referencedWorkflows.length === 0) return;
```

Comment:
Another note to discuss: we shouldn't just return in this case. This would mean a broken project, so we should handle it somehow.